treewide: remove uses of infinite tokens

michoecho opened the pull request:
Refs #7967

We have many redundant ways to express unbounded token intervals. First, there are the infinite tokens `dht::maximum_token()` and `dht::minimum_token()`, which are special variants in `dht::token`. Second, interval types have their own ways of expressing unbounded intervals using `std::optional` bounds. Third, `ring_position` and `ring_position_view` can store positions just after/before the greatest/smallest regular token.

As an extreme example, `nonwrapping_range<ring_position_view>` can be right-unbounded in six different ways: `(x, +inf)` (empty `optional` bound), `(x, maximum_token]`,`(x, ]maximum_token` or `(x, )maximum_token` (right bound after/at/before the infinite `maximum_token`) or `(x, INT64_MAX]`, `(x, ]INT64_MAX` (right bound after/at the greatest possible token).
`nonwrapping_range<ring_position_view>` is uncommon, but `nonwrapping_range<ring_position>` (`partition_range`) is very
common and has 4 different ways to be right-unbounded.

All of this is very confusing. It wouldn't be so bad if all code was prepared to deal with all of those representations, but it seems (but it's very hard to tell - that's the problem) that some places in code expect a particular shape of the interval (usually `(x, y]`), or worse: see a difference between various infinities (like treating `(x, INT64_MAX]` as an single interval, while using `(x, +inf)` only for a half of a wrapping interval?). And there is no way to know if that happens other than studying the code in full detail.

This also results in unnatural corner cases. As an extreme example: `(-inf, )minimum_token`, `(-inf, ]minimum_token`, `(-inf, minimum_token]`, `(-inf, )INT64_MIN+1`, `[minimum_token, INT64_MIN+1)`, `minimum_token[, INT64_MIN+1)`, `minimum_token(, )INT64_MIN+1` are all slightly different possible empty partition intervals, and we need to have them all in mind if we want to check if an interval is empty.

To sort this out, it would be good to remove the redundancy. The idea is to get rid of the infinite tokens, and let infinities be handled by interval and position types. (Even that will leave some redundancy - optional bounds are not needed to express unbounded partition ranges, since positions just after/before the greatest/smallest regular tokens could be used.)
Infinite tokens are particularly bad since they have a triple duty: sometimes they are used to express unbounded ranges, sometimes as a placeholder for an invalid/nonexistent token, sometimes as a position (for example in `db/view`). 

The first step is to remove uses of `dht::maximum_token()` from the codebase. This series removes all uses outside of `dht/` and `test/`. After the remaining few uses in `dht/` are removed, it should be possible to remove `dht::token_kind` (unfortunately it is in IDL, so it will still need to exist at the IDL boundary) and then `dht::token` will be just a wrapper over `int64_t`.

`dht::maximum_token()` is more important to remove since `dht::minimum_token()` has a natural `int64_t` representation, reserved by design. `dht::minimum_token` can remain for the time being to ease the transition (since it's sometimes used as a position).

michoecho commented:
I'm not confident that last few patches don't break anything. See the commit messages. But that exactly the problem - there is no way to know if something doesn't rely on different representations of the same interval.

avikivity commented:
> I'm not confident that last few patches don't break anything. See the commit messages. But that's exactly the problem - there is no way to know if something doesn't rely on different representations of the same interval.

We'll just have to fix any fallout. The biggest risk is that the different representation makes its way across rpc and triggers failures in an old node, where we can't fix anything.

avikivity commented on dht/token.cc:
```diff
@@ -53,6 +55,14 @@ maximum_token() noexcept {
+token greatest_token() noexcept {
+    return greatest_token_static;
+}
+
```
token is so tiny, these could be inlines. I don't mind doing this later.

avikivity replied:
(and no need for static objects, they just pessimize the code, construct everything inline)

michoecho replied:
Right.

avikivity commented on db/size_estimates_virtual_reader.cc:
```diff
@@ -208,10 +208,10 @@ static future<std::vector<token_range>> get_local_ranges(database& db) {
         });
         auto right_inf = boost::find_if(ranges, [] (auto&& r) {
-            return r.start() && (!r.end() || r.end()->value() == dht::maximum_token());
+            return r.start() && (!r.end() || r.end()->value().is_maximum());
```
This can lead to an unintended change: if one of the tokens happened to be the minimum or maximum, we'd misdetect it as left_inf or right_inf.

However, I don't think get_primary_ranges_for() generates ranges with minimum_token - it leads to interval::unwrap() which uses half-open intervals instead. So I think this can just be dropped.

michoecho replied:
I don't understand. This particular diff doesn't change behaviour. `== dht::maximum_token()` does the same thing as `is_maximum()`. The goal of this series so far is to remove `dht::maximum_token` first (which will ensure that we aren't producing +inf tokens, but we may still be receiving them through IDL), and then remove +inf tokens altogether (and then this diff will be revised).

avikivity replied:
Maybe I was taking into account a step that hasn't happened yet, switching maximum_token to greatest_token(). If you don't take that step, and the intent is just to retire maximum_token(), then there is no risk.

avikivity commented on locator/token_metadata.cc:
```diff
@@ -1547,7 +1547,7 @@ void token_metadata_impl::calculate_pending_ranges_for_leaving(
     tlogger.debug("In calculate_pending_ranges: affected_ranges.size={} stars", affected_ranges_size);
     for (const auto& r : affected_ranges) {
-        auto t = r.end() ? r.end()->value() : dht::maximum_token();
+        auto t = r.end() ? r.end()->value() : dht::smallest_token();
```
s/smallest/greatest/

michoecho replied:
No, this is not a mistake. `calculate_natural_endpoints` starts the search (in `token_metadata_impl::_sorted_tokens`) from the passed token, *inclusive*. So passing `maximum_token` actually wraps around and starts from the smallest token. Passing `greatest_token` would start from INT64_MAX, which would change the behaviour if `INT64_MAX` is in `sorted_tokens`.

avikivity replied:
Ah. The wrap-around behavior is so confusing and terrible.

avikivity commented on service/storage_service.cc:
```diff
@@ -3536,7 +3536,7 @@ storage_service::construct_range_to_endpoint_map(
     for (auto r : ranges) {
         res[r] = _db.local().find_keyspace(keyspace).get_replication_strategy().get_natural_endpoints(
-                r.end() ? r.end()->value() : dht::maximum_token());
+                r.end() ? r.end()->value() : dht::smallest_token());
```
greatest?

michoecho replied:
Same answer as above. Changing `maximum_token` to `smallest_token` as the argument to `calculate_natural_endpoints` is a common pattern in this series.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -4670,12 +4674,27 @@ void query_ranges_to_vnodes_generator::process_one_range(size_t n, dht::partitio
+        bool case_1 = e && e->is_minimum();
+        bool case_2 = s && s->is_maximum();
+        // TODO: What about ranges of the form (x, x+1), or (x; x]?
+        // Are they possible here? If so, shouldn't we exclude them here too?
```
It's not good to rely on these things.

avikivity commented:
Another approach is to convert min/max token to half-open intervals. This is safer since it doesn't introduce token numbers that perhaps didn't exist before, but is harder since we need to evaluate the token in a broader context.

(It will also mean we'll need custom deserializers for interval<token> and dht::partition_range)

avikivity commented on service/storage_service.cc:
```diff
@@ -3766,7 +3766,7 @@ storage_service::get_splits(const sstring& ks_name, const sstring& cf_name, rang
-    tokens.push_back(std::move(unwrapped[unwrapped.size() - 1].end().value_or(range_type::bound(dht::maximum_token()))).value());
+    tokens.push_back(std::move(unwrapped[unwrapped.size() - 1].end().value_or(range_type::bound(dht::greatest_token()))).value());
 
     // split_count should be much smaller than number of key samples, to avoid huge sampling error
```
The caller is in thrift:

```c++
    void describe_splits_ex(thrift_fn::function<void(std::vector<CfSplit>  const& _return)> cob, thrift_fn::function<void(::apache::thrift::TDelayedException* _throw)> exn_cob, const std::string& cfName, const std::string& start_token, const std::string& end_token, const int32_t keys_per_split) {
        service_permit permit = obtain_permit();
        with_cob(std::move(cob), std::move(exn_cob), [&]{
            dht::token_range_vector ranges;
            auto tstart = start_token.empty() ? dht::minimum_token() : dht::token::from_sstring(sstring(start_token));
            auto tend = end_token.empty() ? dht::maximum_token() : dht::token::from_sstring(sstring(end_token));
            range<dht::token> r({{ std::move(tstart), false }}, {{ std::move(tend), true }});
```

It's used by clients to attempt to direct queries at the correct node.

michoecho replied:
Why are you pointing this out? Do you mean that this particular place isn't very important?

avikivity replied:
You wrote "I don't understand the full significance of this change." so I'm pointing you at the user.

michoecho commented:
> Another approach is to convert min/max token to half-open intervals. This is safer since it doesn't introduce token numbers that perhaps didn't exist before, but is harder since we need to evaluate the token in a broader context.
> (It will also mean we'll need custom deserializers for interval and dht::partition_range)

Another approach? That was my intention from the beginning, so I don't know what is the first approach :P
I don't think any other approach is safe enough.  Simply coalescing `maximum_token` and `INT64_MAX` or `maximum_token` and `INT64_MIN+1` is not sound, since exclusive bounds are involved (`(x, inf) != (x, INT64_MAX)`). Some custom deserializers will be necessary.

michoecho commented:
> We'll just have to fix any fallout.

Isn't that too optimistic? The `maximum_token` and `minimum_token` stuff is all edge cases – potentially a source of extremely rare bugs (for example, as rare as some partition key being hashed to INT64_MAX). The fallout might be invisible.
This change is taking me awfully long because I'm afraid of off-by-one regressions and I'm trying to study every caller/callee in as much detail as I can to avoid them. Am I too paranoid about this?

avikivity commented:
> > Another approach is to convert min/max token to half-open intervals. This is safer since it doesn't introduce token numbers that perhaps didn't exist before, but is harder since we need to evaluate the token in a broader context.
> > (It will also mean we'll need custom deserializers for interval and dht::partition_range)
> 
> Another approach? That was my intention from the beginning, so I don't know what is the first approach :P
> I don't think any other approach is safe enough. Simply coalescing `maximum_token` and `INT64_MAX` or `maximum_token` and `INT64_MIN+1` is not sound, since exclusive bounds are involved (`(x, inf) != (x, INT64_MAX)`). Some custom deserializers will be necessary.

I assumed you were going to replace those bounds with greatest_token() etc. Looks like I assumed incorrectly. Obviously we agree that the "other" approach is the safe one.

avikivity commented:
> > We'll just have to fix any fallout.
> 
> Isn't that too optimistic? The `maximum_token` and `minimum_token` stuff is all about edge cases – potentially a source of extremely rare bugs (for example, as rare as some partition key being hashed to INT64_MAX). The fallout might be invisible.
> This change is taking me awfully long because I'm afraid of off-by-one regressions and I'm trying to study every caller/callee in as much detail as I can to avoid them. Am I too paranoid about this?

I think we do have tests that cover wraparounds. The risk is indeed from off-by-ones, but  maximum_token() is just another way of specifying a half-open interval, it's not a numeric value. So getting numeric off-by-one here is unlikely.

michoecho commented:
Changes:
- Completed the draft.
- Reverted the introduction of `smallest_token()`. I decided that trying to replace `minimum_token()` (-infinity / INT64_MIN) with `smallest_token()` (INT64_MIN + 1) would do more harm than good. `minimum_token()` is pretty much baked in by design; there is no point trying to avoid it. But it doesn't have to be treated as a special case, as it was until now.

Reviewing guide:
- First, we introduce `greatest_token()` (INT64_MAX - the greatest possible regular token), which we will use as a helper when getting rid of `maximum_token()` (the special +infinity token).
- Patches from `db: size_estimates_virtual_reader: remove uses of maximum_token()` to `cql3: token_restriction: remove use of maximum_token()  remove most ad-hoc uses of `maximum_token()` in the codebase. Those changes are fairly simple: typically, those uses are replaced with `greatest_token()` or empty optional when used in an interval, and `minimum_token()` when used as a position in the ring (usually as an argument to `calculate_natural_endpoints`).
- `dht: token-sharding: add maybe_token_for_next_shard()` to `dht: remove uses of maximum_token() from ring_partition_range_sharder` rewrite `token_for_next_shard()` to return an empty optional instead of `maximum_token()` when there is no more shard boundaries for the given shard. This involves some bigger changes: a rewrite of `ring_partition_range_sharder`, `split_range_to_single_shard()` and `selective_token_range_sharder`.
- `dht: token: remove maximum_token()` to `dht: token: remove token::_kind` get rid of infinite tokens (`maximum_token()` is removed, `minimum_token()` becomes a regular token). Unfortunately, since `token` is an `IDL` type, we still need to deal with `maximum_token()` coming through RPC. We do that by adding custom RPC deserializers for `ring_position` and token intervals. It has to be done at this level, because getting rid of infinite tokens without loss of information requires context. (For example, an exclusive upper bound `(x, maximum_token())` can't be converted to any `(x, y)`, only to `(x, +inf)` of `(x, greatest_token()]`. So the coversion can't happen at the level of `token`.)
This is done in `dht: i_partitioner: add a custom deserializer for ring_position` and `dht: i_partitioner: add custom deserializers for token intervals`. Those patches need special reviewer attention. In particular, please suggest a better place to put this code in. I just stuffed in into a new header (`dht/token_serializers.hh`) and made sure that anything that includes `range.dist.impl.hh` also includes the new header, so that the new template specializations are used, but this is fragile setup.
- `dht: token: merge long_token(), token::to_int64(), and token::raw()` to `dht: token: remove token_kind constructors` are cleanup – removing redundant methods, moving functions between `.cc` and `.hh` and changing pass-by-reference to pass-by-value.

The main result of this series is removing `token_kind` (at least outside of the RPC boundary). Token becomes a wrapper for `int64_t`.

avikivity commented:
I haven't re-reviewed yet, but please add a description of how you tested upgrade interoperability.

btw, the reviewing guide is great, and can be made part of the changelog (in case someone needs to refer to it long after it is committed).

avikivity commented on service/storage_proxy.cc:
```diff
@@ -4676,12 +4680,25 @@ void query_ranges_to_vnodes_generator::process_one_range(size_t n, dht::partitio
-            _i++; // empty range? Move to the next one
-            return;
-        }
+    if (is_singular_range(start, end)) {
```
Perhaps this special case can be removed. But let's not mix it with this change.

avikivity commented on locator/token_metadata.cc:
```diff
@@ -1373,19 +1373,25 @@ dht::token_range_vector token_metadata_impl::get_primary_ranges_for(token right)
+        end = dht::greatest_token();
+        end_inclusive = true;
     }
 
```
I believe this is correct.

avikivity commented on cql3/restrictions/statement_restrictions.cc:
```diff
@@ -600,20 +600,19 @@ dht::partition_range_vector partition_ranges_from_token(const expr::expression&
+    if (bounds.end()) {
+        dht::token t = bounds.end()->value().with_linearized([] (bytes_view bv) { return dht::token::from_bytes(bv); });
+        end = bounds.end()->is_inclusive() ? dht::ring_position::ending_at(t) : dht::ring_position::starting_at(t);
+    }
```
This feeds directly into the changes you made in storage_proxy.

avikivity commented on dht/token.hh:
```diff
@@ -42,6 +42,37 @@ enum class token_kind {
+            _data = std::numeric_limits<int64_t>::min();
+            break;
+        case token_kind::after_all_keys:
+            // Should not be happening. Throw?
```
Why can't this happen? Won't an old node send after_all_keys to a new node, which should translate it to something else? I saw something like this in the ring_position deserializer.

michoecho replied:
The comment was wrong. (It became outdated at some point while I was working on the series, and I forgot to remove it.) Fixed.

avikivity commented on dht/token.hh:
```diff
@@ -36,6 +36,14 @@ namespace dht {
+// However, dht::token is a type used in IDL, so we need to handle token_kind
+// in RPC for backward compatibility. token_kind should only be used on the
+// RPC boundary.
 enum class token_kind {
```
Can we move it to the deserializers?

michoecho replied:
Done.

avikivity commented:
Looks good! Really the only thing missing is upgrade tests (and run some full scans while you have a mixed cluster to exercise ring position range transfers).

michoecho commented on locator/token_metadata.cc:
```diff
@@ -1486,7 +1486,7 @@ void token_metadata_impl::calculate_pending_ranges_for_leaving(
-        auto t = r.end() ? r.end()->value() : dht::maximum_token();
+        auto t = r.end() ? r.end()->value() : dht::minimum_token();
         auto current_endpoints = strategy.calculate_natural_endpoints(t, metadata, can_yield::yes);
         auto new_endpoints = strategy.calculate_natural_endpoints(t, *all_left_metadata, can_yield::yes);
```
Done.

michoecho commented on service/storage_service.cc:
```diff
@@ -2838,8 +2838,8 @@ std::unordered_multimap<dht::token_range, inet_address> storage_service::get_cha
-        auto end_token = r.end() ? r.end()->value() : dht::maximum_token();
-        auto new_replica_endpoints = ks.get_replication_strategy().calculate_natural_endpoints(end_token, temp, utils::can_yield::yes);
+        auto search_token = r.end() ? r.end()->value() : dht::minimum_token();
+        auto new_replica_endpoints = ks.get_replication_strategy().calculate_natural_endpoints(search_token, temp, utils::can_yield::yes);
```
Done.

michoecho commented on dht/token_serializers.hh:
```diff
@@ -0,0 +1,87 @@
+            size_type size = deserialize(buf, boost::type<size_type>());
+            auto in = buf.read_substream(size - sizeof(size_type));
+
+            auto legacy_token = deserialize(in, boost::type<dht::legacy_token>());
```
It is retained, my comment earlier was wrong.

michoecho commented:
v2:
- Rebased.
- Some very minor changes (corrected a commit message typo, changed a pass-by-reference to pass-by-value, etc).
- Expanded commit messages where requested.
- Moved `token_kind` and `legacy_token` to `token_serializers.hh`.

michoecho commented:
> please add a description of how you tested upgrade interoperability

I have run `upgrade_test.py` and `rolling_upgrade_test.py` from `scylla-dtest`. (I think I have run them correctly, because they found a bug which slipped through the regular unit tests and gating dtests). Is that enough?

> run some full scans while you have a mixed cluster to exercise ring position range transfers

How would that "exercise ring position range transfers"? Can you elaborate?

michoecho commented:
@avikivity ^

avikivity commented:
> > please add a description of how you tested upgrade interoperability
> 
> I have run `upgrade_test.py` and `rolling_upgrade_test.py` from `scylla-dtest`. (I think I have run them correctly, because they found a bug which slipped through the regular unit tests and gating dtests). Is that enough?


I don't know what they do. From which version did you upgrade? Do they run a workload during the upgrade?

Likely it is sufficient.

> 
> > run some full scans while you have a mixed cluster to exercise ring position range transfers
> 
> How would that "exercise ring position range transfers"? Can you elaborate?


It sends non-singular interval<ring_position>.

michoecho commented:
> I don't know what they do. Do they run a workload during the upgrade?

`upgrade_test.py` (cluster size = 3): 

```
    def test_one_node_upgrade(self):
        """
        Test upgrade one node.
        1. Prefill the table before upgrade
        2. Upgrade one node
        3. Validate the data is not corrupted
        4. Add new data and validate
        """
        ...

    def test_cluster_upgrade(self):
        """
        Test upgrade all nodes in the cluster sequentially.
        Prefill the table before upgrade and validate the data is not corrupted
        """
        ...
```

`rolling_upgrade_test.py` is similar, but also tests version rollback.

> From which version did you upgrade?

From 4.5. Should I test older versions too?

avikivity commented:
> > I don't know what they do. Do they run a workload during the upgrade?
> 
> `upgrade_test.py` (cluster size = 3):
> 
> ```
>     def test_one_node_upgrade(self):
>         """
>         Test upgrade one node.
>         1. Prefill the table before upgrade
>         2. Upgrade one node
>         3. Validate the data is not corrupted
>         4. Add new data and validate
>         """
>         ...
> 
>     def test_cluster_upgrade(self):
>         """
>         Test upgrade all nodes in the cluster sequentially.
>         Prefill the table before upgrade and validate the data is not corrupted
>         """
>         ...
> ```


Looks like it's doing too little. I want RPC from old->new and new->old to be happening with a mixed version cluster.

> 
> `rolling_upgrade_test.py` is similar, but also tests version rollback.
> 
> > From which version did you upgrade?
> 
> From 4.5. Should I test older versions too?

No. We only promise one minor version back, and there are no material differences in passing ring_positions and tokens between 4.5 and older versions.

michoecho commented:
> Looks like it's doing too little. I want RPC from old->new and new->old to be happening with a mixed version cluster.

I have modified that test to do the following:
1. Setup a 3-node cluster (with version 4.5).
2. Insert 100 new rows.
3. Upgrade a node to the new version.
4. Check that full scan returns all data inserted so far.
5. for _ in range(10): for coordinator in nodes: insert 100 rows and check that full scan returns all data inserted so far

Good enough?

michoecho commented:
@avikivity Bump.

avikivity commented:
Sorry for the delay. 3 nodes may be too little if RF=3, since it means all nodes have all data, and the coordinator may collapse all the token intervals into one. Combined that with having just 100 rows, we end up with the entire scan being just one read, and perhaps we miss something. I'd like to see a test with RF<cluster_size, to make sure those reads don't get short-circuited by those optimizations.

michoecho commented:
> @michoecho was this PR abandoned?

@denesb Inadvertently. I had to switch to something else while waiting for review, some merge conflicts happened in the meantime, and here we are.
It should be still usable/revivable, though. The number of conflicts is nowhere near as high as it appears (most of it comes from 152f1d017e2d3c9933020f421361d05d717088f4, which is largely independent from everything else).

denesb commented:
Looking at the comments it was really close to getting in. It'd be a shame to let it rot. Of course we need help from @avikivity to move it forward as he was the main reviewer.

bhalevy commented:
ping, please rebase and let's revive review and hopefully get it merged

mykaul commented:
@michoecho , @bhalevy - is this item still being worked on?

michoecho commented:
> @michoecho , @bhalevy - is this item still being worked on?

Are you asking whether it should be closed?
I would rather leave it open. I think we want to apply this eventually. But it's something that should wait until we solve our more important problems.
